Join hints

lutovich opened the pull request:
This PR introduces join hints in Cypher for greedy and IDP planners.
Hints could be used with `USING JOIN ON {node identifier}` syntax.
Rule planner simply ignores join hints.
Hints applied to patterns with less than 2 relationships result in exception.

Example:

Query: `MATCH (a)-->(m)<--(z) RETURN m`
Plan:

```
+ProduceResults
|              
+Projection      
|              
+Filter        
|              
+Expand(All)   
|              
+Expand(All)   
|              
+AllNodesScan  
```

Query: `MATCH (a)-->(m)<--(z) USING JOIN ON m RETURN m`
Plan:

```
+ProduceResults
|              
+Projection      
|              
+Filter        
|              
+NodeHashJoin  
|\             
| +Expand(All) 
| |            
| +AllNodesScan
|              
+Expand(All)   
|              
+AllNodesScan  
```

It also adds a randomised test that generated different patterns for `MATCH`, executes query with such pattern & join hint and asserts that hash join is planned.

lutovich committed:
Parser support for join hints

lutovich committed:
Greedy and IDP planners support for join hints

This commit makes both planners to treat queries with solved join hints
as valuable thus making them appear as planning outcome.

Rule planner is changed to skip join hints.

Added a scalacheck randomized integration test for join hints
in both planners.

lutovich committed:
Make join hints work for single hop patterns

Commits adds possibility to plan hash joins for patterns like (a)-->(b).
Such joins would be planned on top of expands.
In reality, join on top of expand is probably not efficient so it would never
picked up by cost model but could be forced by a join hint.

lutovich commented:
@davidegrohmann all mentioned issues have been fixed. I've squashed commits so comments disappeared.
@systay would be great if you could review this PR.

systay commented on community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/convert/commands/StatementConverters.scala:
```diff
@@ -230,6 +230,8 @@ object StatementConverters {
       case ast.UsingScanHint(identifier, label) =>
         commands.NodeByLabel(identifier.name, label.name)
+      case ast.UsingJoinHint(identifier) =>
+        commands.NodeJoin(identifier.name)
```
Since we are simply going to be throwing out the join hint for the rule planner, we might as well not create commands objects representing the hint, and instead just ignoring it here instead. Remember to send a notification when the hint is being dropped though

lutovich replied:
Fixed. Now only notification is logged here.

lutovich committed:
Fixed warnings for join hints in RULE planner

Previously join hints were converted to commands, included into query and then
immediately filtered out during planning without logging a warning. This commit
makes it so that join hints would not get converted into commands and would
not appear in query. Warning is logged when ast -> command conversion
attempt is made.

lutovich commented:
@systay I've fixed warnings for rule planner and added var-length to `JoinHintPlanningIntegrationTest`.

I did not fix throwing exceptions instead of warnings because currently only `IndexHintUnfulfillableNotification` can result in exception. This is strange and should probably be revisited in a separate PR.

dhimmel commented:
I am struggling to understand what `USING JOIN ON` does. In the example below, I expected the planner to start the search from both ends and join at `n2`. My intent was to specify a single `NodeHashJoin` to occur on `n2`.

``` cypher
PROFILE
MATCH paths = (n0:Compound)-[:CAUSATION_CcSE]-(n1)-[:CAUSATION_CcSE]-(n2)-[:DOWNREGULATION_CdG]-(n3)-[:UPREGULATION_DuG]-(n4:Disease)
USING INDEX n0:Compound(identifier)
USING INDEX n4:Disease(identifier)
USING JOIN ON n2
WHERE n0.identifier = 'DB00014'
AND n4.identifier = 'DOID:1324'
RETURN COUNT(paths) AS PC
```

However, on 2.3.1 the planner performs several `NodeHashJoin`s and `AllNodesScan`s

![plan](https://cloud.githubusercontent.com/assets/1117703/11602727/5a288a6c-9a91-11e5-8714-f5ef6765ecf1.png)

lutovich commented:
@dhimmel hey, there is a recent PR that could possible help with this problem: https://github.com/neo4j/neo4j/pull/5927. I can't say for sure because I'm lacking context a bit.
@henriknyman could you please take a look?

lutovich commented:
@dhimmel on recent 2.3 dev build plan indeed looks much better:
![plan](https://cloud.githubusercontent.com/assets/1638145/11603452/1427f34e-9ae2-11e5-854a-2a690ce79d4d.png)

so it does two index lockups for `n0` and `n4` then expands from both ends towards `n2` and performs a hash join.
This fix is available in 3.0.0-M01 release and would be part of 2.3.2.

dhimmel commented:
Thanks @lutovich. It's exciting to have an unexpected behavior result from a bug rather than user error! Congrats on the [3.0 milestone](http://neo4j.com/blog/neo4j-3-0-milestone-1-release/) -- can't wait for bolt and the python driver!
